# シンプルなニューラルネットワーククラスの実装
* ここでは、シンプルなニューラルネットワーククラスを実装する。

In [1]:
import numpy as np
from common.activations import softmax
from common.grad import numerical_gradient
from common.loss import cross_entropy_error

* 「ニューラルネットワークにおいて勾配を求める」とは、損失$L$の$W$に対する偏微分$\displaystyle \frac{\partial L}{\partial {\bf W}}$を求めるということである。  
    例、  
    ${\bf W}=\begin{pmatrix} w_{11} & w_{12} & w_{13} \\ w_{21} & w_{22} & w_{23} \end{pmatrix}$  
  
    $\displaystyle \frac{\partial L}{\partial {\bf W}}=\begin{pmatrix} \frac{\partial L}{\partial {w_{11}}} &
                                                                                      \frac{\partial L}{\partial {w_{12}}} &
                                                                                      \frac{\partial L}{\partial {w_{13}}} \\
                                                                                      \frac{\partial L}{\partial {w_{21}}} &
                                                                                      \frac{\partial L}{\partial {w_{22}}} &
                                                                                      \frac{\partial L}{\partial {w_{23}}}
                                                                                      \end{pmatrix}$  
                                                                                            
  
    
* $\displaystyle \frac{\partial L}{\partial {\bf W}}$は、${\bf W}$のそれぞれの要素が微小量変化したときの$L$の変化量を表す。


### [演習]
* シンプルなニューラルネットワークにおいて、予測、損失、勾配を計算する以下のクラスを完成させましょう


In [2]:
# ヒント
np.random.seed(seed=1234)
print(np.random.randn(2,3)) # 標準正規分布からランダムにサンプリング

[[ 0.47143516 -1.19097569  1.43270697]
 [-0.3126519  -0.72058873  0.88716294]]


In [3]:
class simpleNet:
    def __init__(self, seed=1234):
        np.random.seed(seed)
        self.W = np.random.randn(2,3)#Wの初期値
        
    def predict(self, x):
        """
        予測関数
        """
        z = np.dot(x, self.W)
        y = softmax(z)
        return y
    
    def loss(self, x, t):
        """
        損失関数
        """
        y = self.predict(x)
        loss = cross_entropy_error(y, t)
        return loss
    
    def gradient(self, x, t):
        """
        勾配計算関数
        """
        grads={}
        f = self.loss
        W = self.W
        grads["W"] = numerical_gradient(f, x, W, t)

        return grads

In [4]:
# 重みの初期値の確認
snet = simpleNet()        
print("snet.W=")
print(snet.W)
print()

# 予測関数の確認
x = np.array([0.6, 0.9])
y = snet.predict(x)
print("y=", y)
print()
print("np.argmax(y)=", np.argmax(y))
print()

# 損失を求める関数の確認
t =np.array([0, 0, 1]) #正解ラベル
print("loss=",snet.loss(x, t))
print()

# 勾配を求める
grad = snet.gradient(x, t)
print("grad=")
print(grad)

net.W=
[[ 0.47143516 -1.19097569  1.43270697]
 [-0.3126519  -0.72058873  0.88716294]]

y= [0.15391894 0.03932405 0.80675701]

np.argmax(y)= 2

loss= 0.21473263744557

grad=
{'W': array([[ 0.09235135,  0.02359443, -0.11594578],
       [ 0.13852703,  0.03539164, -0.17391867]])}


### [gradの解釈]
* 求められたgradは、例えば、次のように解釈できる
    * $w_{11}$を正の方向に微小量$h$だけ変化させたとき、$L$は約0.092増える。
    * $w_{13}$を負の方向に微小量$h$だけ変化させたとき、$L$は約0.116増える。
    * $L$は$0$に近づけたいので、$w_{11}$は負の方向に更新し、$w_{13}$は正の方向に更新するのが良い、となる。

### [演習]
* 重みWの初期値を変えてみましょう。